src/libraries/System.Net.NameResolution/src/System/Net/Dns.cs

```diff
@@ -666,7 +664,18 @@ private static Task<TResult> RunAsync<TResult>(Func<object, long, TResult> func,
                     Debug.Assert(!Monitor.IsEntered(s_tasks));
                     try
                     {
-                        return func(key, startingTimestamp);
+                        using (cancellationToken.UnsafeRegister(static s =>
+                        {
+                            // In case a request is cancelled, start a new queue for this key.
+                            // This helps avoiding congestion when a getaddrinfo call runs for too long.
```

Comment:
What's the rationale for why a cancellation request means we no longer want to serialize?

Reply:
One of the original motivations for doing this queueuing was a DNS request that could get stalled, and thus all requests for that DNS endpoint would end up blocking all threads in the pool.  Doesn't this effectively put us back into the same place?

Reply:
From what I see, #48566 doesn't talk about stalled (up to several seconds) requests but about slow (tens of milliseconds?) requests using up the ThreadPool threads. This does not remove the serialization of those. The queue is only being reset when a cancellation kicks in which is typically triggered by a `ConnectTimeout`, which is usually in a multiple seconds range. If the customer mentioned in #34633 had stalled requests, they would have likely came back to us reporting #81023 much earlier.

cc @davidfow if you can recall how long did the slow requests take in #34633.  Without that I would assume it's a case similar to what we discussed in https://github.com/dotnet/runtime/pull/92862#discussion_r1342186061 and "slow" means that the requests are in the 5-50 ms range.

Reply:
Right, so if every request for a given destination ends up taking a longer period of time and ends up being canceled as a result, every cancellation is going to cause subsequent requests to run concurrently with ones already there, and if it happens enough, won't it similarly spread out over all of the pool threads?

> when a cancellation kicks in which is typically triggered by a ConnectTimeout

That's one source. The Dns APIs that support cancellation are public.

Reply:
> if every request for a given destination ends up taking a longer period of time and ends up being canceled as a result, every cancellation is going to cause subsequent requests to run concurrently with ones already there, and if it happens enough, won't it similarly spread out over all of the pool threads?

Yes, but how likely is this scenario, and is it more probable than hitting #81023? To me this looks like having all requests to the same host stalled for a long enough period to starve the ThreadPool is a catsrophic event which would have serious impact on the application's funcionality even without the pool starvation, and the customer has to go and fix the environmental setup anyways. Occasional stalled requests are more common IMO, and the problem is that they are currently killing a possibly long chain of follow-up requests.

> That's one source. The Dns APIs that support cancellation are public.

IMO this doesn't really change things. Regularly occuring cases of cancellation are be most likely timeouts, and cancelling a significant percentage of DNS lookups by setting an overly short timeout doesn't look like a sane scenario to me.

Reply:
> Occasional stalled requests are more common IMO

What is the cause of the stalled requests you're seeing that makes this the case?

Reply:
In #92054 we see the OS resolver (succesfully) retrying a query after the default `timeout` of 5 seconds, and `getaddrinfo` is blocking for this 5sec period. The retries are sporadic and do not come in bursts. They are caused by a packet loss in my understanding, which shouldn't be very uncommon.

I don't have such confident information from the other cases, because this is typcially happening in production systems in a sporadic manner. Currently I'm working with @matt-augustine checking if a private build of the changes in this PR helps with [their symptoms](https://github.com/dotnet/runtime/issues/81023#issuecomment-1730141380). Would that bring enough confidence for you that the change is net positive?